algorithms/kernel/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i

```diff
@@ -0,0 +1,933 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Implementation of auxiliary functions for decision forest classification
+//  hist method.
+//--
+*/
+
+#ifndef __DF_CLASSIFICATION_TRAIN_HIST_ONEAPI_IMPL_I__
+#define __DF_CLASSIFICATION_TRAIN_HIST_ONEAPI_IMPL_I__
+
+#include "algorithms/kernel/dtrees/forest/classification/oneapi/df_classification_train_hist_kernel_oneapi.h"
+#include "algorithms/kernel/dtrees/forest/classification/oneapi/cl_kernels/df_batch_classification_kernels.cl"
+
+#include "algorithms/kernel/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i"
+#include "algorithms/kernel/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.i"
+#include "algorithms/kernel/dtrees/forest/classification/df_classification_model_impl.h"
+#include "algorithms/kernel/dtrees/forest/classification/oneapi/df_classification_tree_helper_impl.i"
+
+#include "externals/service_ittnotify.h"
+#include "externals/service_rng.h"
+#include "externals/service_math.h" //will remove after migrating finalize MDA to GPU
+#include "services/buffer.h"
+#include "data_management/data/numeric_table.h"
+#include "services/env_detect.h"
+#include "services/error_indexes.h"
+#include "service/kernel/service_data_utils.h"
+#include "service/kernel/service_algo_utils.h"
+#include "service/kernel/service_arrays.h"
+#include "service/kernel/service_utils.h"
+#include "algorithms/kernel/engines/engine_types_internal.h"
+#include "oneapi/internal/types.h"
+
+using namespace daal::algorithms::decision_forest::internal;
+using namespace daal::algorithms::decision_forest::classification::internal;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace decision_forest
+{
+namespace classification
+{
+namespace training
+{
+namespace internal
+{
+template <typename algorithmFPType>
+static services::String getFPTypeAccuracy()
+{
+    if (IsSameType<algorithmFPType, float>::value)
+    {
+        return services::String(" -D algorithmFPTypeAccuracy=(float)1e-5 ");
+    }
+    if (IsSameType<algorithmFPType, double>::value)
+    {
+        return services::String(" -D algorithmFPTypeAccuracy=(double)1e-10 ");
+    }
+    return services::String();
+}
+
+inline char * utoa(uint32_t value, char * buffer, size_t buffer_size)
```

Comment:
This also looks like a service functionality.

Reply:
Yes will migrate it to service for next release